src/JuMP.jl

```diff
@@ -102,7 +102,8 @@ the constructor call `IpoptOptimizer(print_level=0)`:
 with_optimizer(IpoptOptimizer, print_level=0)
 ```
 """
-function with_optimizer(constructor::Type, args...; kwargs...)
+function with_optimizer(constructor::Union{Function, DataType, UnionAll},
```

Comment:
The signature should match the docstring. Either drop the types from the `constructor` argument or add them to the docstring. I don't think the types add much since the compiler can't specialize, but we do have to be aware of postponing a `MethodError` that may happen in `(optimizer_factory::OptimizerFactory)()`. Maybe an explicit `hasmethod` in `(optimizer_factory::OptimizerFactory)()` with a nice error message would address that?